ncm-systemd/src/main/perl/Systemd/Service.pm

```diff
@@ -619,16 +619,26 @@ sub change
         }
     }
 
-    # TODO: same TODOs as with states
-    foreach my $act (sort keys %$acts) {
-        my @units = @{$acts->{$act}};
-
-        # TODO: process units wrt dependencies?
-        if (@units) {
-            # TODO: trap exitcode and stop any further processing in case of error?
-            # CAF::Process logger is sufficient
-            systemctl_command_units($self, $change_activation->{$act}, @units);
+    my $multi_user_active = systemctl_is_active($self, "multi-user.target");
+
+    # Avoid starting/stopping services during boot and shutdown. Trying to stop
+    # services which are already queued to be stopped may cause deadlocks.
+    # Trying to start services which were just stopped may lead to unexpected
+    # behavior and data loss.
+    if ($multi_user_active eq 'active') {
+        # TODO: same TODOs as with states
+        foreach my $act (sort keys %$acts) {
+            my @units = @{$acts->{$act}};
+
+            # TODO: process units wrt dependencies?
+            if (@units) {
+                # TODO: trap exitcode and stop any further processing in case of error?
+                # CAF::Process logger is sufficient
+                systemctl_command_units($self, $change_activation->{$act}, @units);
+            }
         }
+    } else {
+        $self->info("multi-user.target is not active, not starting or stopping any services.");
```

Comment:
would it make sense to eg wait and recheck a few times, instead of doing nothing? in case a boot, it could run later whne multi-user target is active.

i guess this should also return an error (since nothing was done). and ncm-ncd should be rerun

Reply:
During shutdown, waiting and retrying would definitely not do any good. I'm less sure about boot time, but I would still be hesitant to interfere with the boot sequence, because experience tells us dependencies are not always correct and/or complete.

If you really want to make sure services which were changed to be enabled while the host was down are started after reboot, it would be possible to schedule `ncm-ncd --configure systemd` to run _after_ `multi-user.target` has been reached.